pkg/gitversion/gitversion.go

```diff
@@ -365,21 +354,56 @@ func mostRecentTag(repo *git.Repository, ref plumbing.Hash, isPrerelease bool,
 	}
 
 	var mostRecentTag *plumbing.Reference
-	walker := object.NewCommitPreorderIter(commit, nil, nil)
-	err = walker.ForEach(func(commit *object.Commit) error {
-		isExact, exact, err := isExactTag(repo, commit.Hash, isPrerelease, tagFilter)
-		if err != nil {
-			return err
-		}
+	walk := func(tags func() (storer.ReferenceIter, error)) func(*object.Commit) error {
+		return func(commit *object.Commit) error {
+			tags, err := tags()
+			if err != nil {
+				return fmt.Errorf("resolving repo tags: %w", err)
+			}
+			isExact, exact, err := isExactTag(tags, commit.Hash, isPrerelease, tagFilter)
+			if err != nil {
+				return err
+			}
+
+			if !isExact {
+				return nil
+			}
+
+			mostRecentTag = exact
+			return storer.ErrStop
 
-		if !isExact {
-			return nil
 		}
+	}
 
-		mostRecentTag = exact
-		return storer.ErrStop
-	})
+	// Walk the commits that are checked out, in order.
+	walker := object.NewCommitPreorderIter(commit, nil, nil)
+	err = walker.ForEach(walk(repo.Tags))
+	if err != nil && !errors.Is(err, plumbing.ErrObjectNotFound) {
+		return false, nil, fmt.Errorf("could not find local object: %w", err)
+	}
+	// We have found a local tag that is a parent, so return it.
```

Comment:
Curious is this behavior what pulumi build system needs? This would prefer a local tag. Suppose you forgot to do fetch tags. So origin has more recent tags than local. For example:

```
local: c1=v0.9 --> c2 --> c3=HEAD
origin: c1=v0.9 --> c2 --> c3=v1.0
```

Current algo will decide we're at 0.9-pre.something

If you had this though:
 
```
local: c1 --> c2 --> c3=HEAD
origin: c1=v0.9 --> c2 --> c3=v1.0
```

The also would decide we're at v1.0?

Reply:
In practice, I think the only way for `pulumictl` to see a local and remote history as in your example (👇) is on a dev machine that has not fetched tags in a while.

```
local: c1 --> c2 --> c3=HEAD
origin: c1=v0.9 --> c2 --> c3=v1.0
```

I can change when we execute a remote search to only trigger when we are missing history, which would prevent this corner case.

Reply:
To step back why is it consulting both local and remote and tries to merge it? Can it just look in remote?

Reply:
Hypothesis.

The operation you need in CI builds is this. Fetch me the bare --depth 1 commit `X` from remote to get the code, and compute the version. To compute the version you only need to look in remote/origin. 

The operation you need locally is this. I have a full non-bare repo and I need to compute version based on local tags ONLY. If I introduce tags I want it to see them. It's my job as dev to sync tags with remotes and so forth.

Can these two use cases perhaps be separate commands instead of coding up merging remote and local?

Reply:
I'll add a `--remote={yes,no}` flag that defaults to `no`. This will let us:
1. Enable this behavior separate from a merge to master.
2. Disable it by for local builds.